### PR TITLE
Refine lazy-loading implementation by not lazy-loading elements that are likely to appear above the fold

### PR DESCRIPTION
**Pull request by felixarntz**
This PR implements the refinement described in [this post](https://make.wordpress.org/core/2021/07/15/refining-wordpress-cores-lazy-loading-implementation/) as a core enhancement.

* Introduce a new function `wp_get_loading_attr_default( $context )` that should be used in place of the hard-coded 'lazy' string defaults used throughout the codebase so far.
    * The function still returns 'lazy' in most circumstances.
    * However, for contexts `the_content` and `the_post_thumbnail` (new, see below) it returns `false` (to omit the `loading` attribute) for the very first image or iframe within the main query loop.
* Introduce a new filter `wp_omit_loading_attr_threshold` in order to allow fine tuning of the default threshold of **1** content image or iframe (e.g. based on special layout usage by a theme). For example, if the filter is used to set the threshold to 3, the first 3 images or iframes won't be lazy-loaded.
* A new context `the_post_thumbnail` is introduced, used within `get_the_post_thumbnail()` and thus taking precedence over the `wp_get_attachment_image` context that would be used by `wp_get_attachment_image()`.
* One rather minor, but significant change has been made to `wp_filter_content_tags()`:
    * The function used to first go over all images and then all iframes.
    * However, since now sequence truly matters in order to decide whether or not to add a `loading` attribute, the function now goes over all the matches in their correct order. Nothing else has changed in the logic there, and no existing tests are affected by this change.
* Test coverage is present.

Trac ticket: https://core.trac.wordpress.org/ticket/53675

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by mtias** [comment]
@felixarntz have you thought about how we could have more precise heuristics going forwards that can take the semantics and structure of blocks into account to get a sense for what is actually deferrable? For example, an image block or a site logo used in a header template part would be strong indicatives of being above the fold. "The first image of the content" seems instead like a rudimentary measure, that varies a lot depending on preceding layout. With block themes we should have some ahead-of-time awareness of layout which we can use to produce more meaningful instructions.

**Comment by felixarntz** [comment]
@mtias Great point, there is definitely room for further refinement as we're moving towards a world of block themes. I've been thinking about detecting the header template part as well.

The analysis I've conducted using the fix here (even though technically I agree that "the first content image" _sounds_ rudimentary) has proven that it already significantly improves performance across the board, so I think it's a useful enhancement on its own - I don't think we should overcomplicate it for now. Once there are more block-based themes available, I agree it would make sense to refine further by using the semantics of blocks. For example, we could also make decisions like "if the first block is a gallery block with three images per row, omit lazy-loading on the first _three_ images".

**Comment by westonruter** [comment]
Also wanted to note that I found that [3,801 themes](https://wpdirectory.net/search/01FDAKAH3ZB1A7PNTP1W6RRZ30) (out of 4,145)  make use of the `main` element. I was surprised (and heartened) to see that 90%+ of themes use this semantic element. For the AMP plugin, we've used this as an easy way to identify all images that appear in the header, by [looking for all that occur before the `main`](https://github.com/ampproject/amp-wp/blob/24923a403e6c010228b143375309336509f2eda1/src/Optimizer/Transformer/DetermineHeroImages.php#L120-L135). Granted, this wouldn't work the same way in core since the DOM of the page is not available, but eventually in a FSE context this could be implemented similarly through filtering of blocks.

File: src/wp-includes/media.php [hunk 1]
```diff
@@ -5176,3 +5190,67 @@ function wp_get_webp_info( $filename ) {
...
+	 *
+	 * @param int $threshold Threshold for how many of the first content media elements to not lazy-load. Default 1.
+	 */
+	$eager_threshold = apply_filters( 'wp_omit_loading_attr_threshold', 1 );
+
+	// Increase the counter since this is a main query content element.
+	if ( ! isset( $wp_content_media_count ) ) {
+		$wp_content_media_count = 0;
+	}
+	$wp_content_media_count++;
```
Comment (westonruter) [inline src/wp-includes/media.php:5247]: One complication with this approach: if a plugin happens to `apply_filters('the_content', get_post()->post_content)` before also doing `the_content()` (e.g. for analysis), then this could result in counts that are “thrown away”. 

I've run into this problem where I was attempting to inline styles for the first occurence of a block only to find on a site with a table of contents plugin that the styles don't show up because they were injected while applying `the_content` filters without rendering the output: https://github.com/westonruter/syntax-highlighting-code-block/issues/286.

See problematic code from that TOC plugin: https://github.com/openlab-at-city-tech/openlab/blob/cacf07eceb2addc1f431c6ffeae4e34b322c1899/wp-content/plugins/fixed-toc/frontend/class-frontend-control.php#L138-L142

File: src/wp-includes/post-thumbnail-template.php [hunk 2]
```diff
@@ -176,6 +176,19 @@ function get_the_post_thumbnail( $post = null, $size = 'post-thumbnail', $attr =
...
+		// Get the 'loading' attribute value to use as default, taking precedence over the default from
+		// `wp_get_attachment_image()`.
+		$loading = wp_get_loading_attr_default( 'the_post_thumbnail' );
+
+		// Add the default to the given attributes unless they already include a 'loading' directive.
+		if ( empty( $attr ) ) {
+			$attr = array( 'loading' => $loading );
+		} elseif ( is_array( $attr ) && ! array_key_exists( 'loading', $attr ) ) {
+			$attr['loading'] = $loading;
+		} elseif ( is_string( $attr ) && ! preg_match( '/(\?|&)loading=', $attr ) ) {
```
Comment (westonruter) [inline src/wp-includes/post-thumbnail-template.php:188]: The `?` here doesn't seem right. The `$attr` here is a query string, not a URL with a query string attached to it. So I think this is rather needing to see if `loading=` occurs at the beginning of the query string or as a key somewhere inside:

```suggestion
		} elseif ( is_string( $attr ) && ! preg_match( '/(^|&)loading=', $attr ) ) {
```

File: src/wp-includes/media.php (hunk 1 above)
Comment (westonruter) [reply src/wp-includes/media.php:5247]: I suppose the worst case is that here is that more lazy-loaded images get served at the beginning of the content, which wouldn't be any worse than they are today. So probably not a big deal, but it is an edge case.

**Comment by mtias** [comment]
@felixarntz yes, I didn't mean rudimentary in a disparaging way, just that it feels we could be doing more tailored targeting there with block semantics.

**Review by adamsilverstein — COMMENTED** [review]
Overall this looks good to me. It would be good to explain in the dev notes why `wp_omit_loading_attr_threshold` might be useful. 

Thinking out loud: it almost feels like lazy should be part of src set - some images might be in-viewport on large screens and out-of-viewport on smaller screens, if would be nice if the `loading` attribute could be applied with this type of media query.

**Comment by adamsilverstein** [comment]
Testing with `twentynineteen`, the header/top image seems to still get the lazy attribute, while for content's first image which you have to scroll to reach does not get it.

image

**Comment by adamsilverstein** [comment]
verified working correctly in `twentytwenty` so must be something specific about how 2019 outputs that top image.

File: src/wp-includes/media.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-includes/media.php:5247]: @westonruter That's a fair point overall, but the code here has additional checks in place to make it less likely (while certainly not impossible) that that happens. For example, the TOC plugin you referenced runs that logic on `wp_enqueue_scripts`. With the code here, that wouldn't result in problems because it only applies when `in_the_loop()`, among other things (see the beginning of this function).

File: src/wp-includes/post-thumbnail-template.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/post-thumbnail-template.php:188]: Good catch, updated!

**Comment by felixarntz** [comment]
@adamsilverstein 
> Testing with `twentynineteen`, the header/top image seems to still get the lazy attribute, while for content's first image which you have to scroll to reach does not get it.

That was expected behavior with the proposed fix, since that theme is rendering the featured image outside of the loop. It raises a good point though that maybe we should support that case, since it is fairly common to render the featured image on top of the page when `is_singular()`. I've implemented this in https://github.com/WordPress/wordpress-develop/pull/1580/commits/03a95a93ecb0b267533e8bfb5825e690541fe152. Similar to the other changes in this PR, this is not 100% reliable, but it won't have adverse effects, and I'm assuming it will bring even better LCP results than without it.

I will run a separate follow-up A/B test to compare. Until then, happy to hear your feedback @westonruter on this new commit as well.

**Comment by felixarntz** [comment]
@adamsilverstein @westonruter @thierryA I've ran a test between both versions of the fix (see above https://github.com/WordPress/wordpress-develop/pull/1580#issuecomment-931592700) across the 30 most "popular" themes based on the wordpress.org API. To clarify, the two different versions are:
* "corefix1" counts all `the_content` and `the_post_thumbnail` images as long as they're also `in_the_loop()`.
* "corefix2" is based on the latest iteration here, so it supports one extra condition where `the_post_thumbnail` is also counted for a "singular" context, even if not `in_the_loop()` (where it e.g. covers the special behavior in Twenty Nineteen).

The difference between the two is barely noticeable:
* For both LCP and image KB loaded, the median difference is 0%.
* Especially for LCP there are no notable wins on either end, although the number of tests where there is an improvement in "corefix2" is slightly higher than the number of tests with an improvement in "corefix1". That isn't statistically sufficient for a decision though.
* In image KB, there is barely any difference, in most individual tests the difference is 0%. It's in fact only 4 themes where any scenario gave a result other than 4%, and two of those are indeed Twenty Nineteen and Twenty Seventeen, which both for a singular post have that special behavior that they display the featured image outside of the loop.

Given that there is no win across the board from also counting featured images in singular context outside the loop and since it is also a special case anyway, I suggest we revert https://github.com/WordPress/wordpress-develop/pull/1580/commits/03a95a93ecb0b267533e8bfb5825e690541fe152 again and go with the original approach. I don't think adding a special case just for some core themes with a non-standard implementation is justifiable. Thoughts?

**Comment by ThierryA** [comment]
> Given that there is no win across the board from also counting featured images in singular context outside the loop and since it is also a special case anyway, I suggest we revert 03a95a9 again and go with the original approach. I don't think adding a special case just for some core themes with a non-standard implementation is justifiable. Thoughts?

Sounds totally reasonable to me, +1!

File: src/wp-includes/media.php [hunk 3]
```diff
@@ -5177,3 +5191,67 @@ function wp_get_webp_info( $filename ) {
...
+ *
+ * @global int $wp_content_media_count Content media count used to enumerate elements for the contexts 'the_content'
+ *                                     and 'the_post_thumbnail'.
+ *
+ * @param string $context Context for the element for which the `loading` attribute value is requested.
+ * @return string|bool The default `loading` attribute value. Either 'lazy', 'eager', or a boolean `false`, to indicate
+ *                     that the `loading` attribute should be skipped.
+ */
+function wp_get_loading_attr_default( $context ) {
+	global $wp_content_media_count;
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply src/wp-includes/media.php:5220]: Yeah, I guess it may be fine to actually remove the awareness of `$wp_content_media_count` from the tests entirely. They may be a bit harder to follow when reading the tests, but some extra doc comments could help with that. The tests would still already test whether the correct value for the `loading` attribute is returned so it would be sufficient.

I'll update it to use a static, then we can't influence it from the tests, but per the above I think that's okay.

[the remark this answers is not in the document]
Comment (felixarntz) [reply src/wp-includes/media.php]: I've introduced a simple function `wp_increase_content_media_count()` to handle the static count. Using just a static won't work as it cannot be influenced from the outside at all. And while we don't want it to be typically modified from the outside, preventing it entirely is not a good idea as it also prevents us from being able to test it.

I think using a function is an elegant approach that allows us to control the value while not making it too easy to e.g. accidentally override/modify it.

[the remark this answers is not in the document]
Comment (azaozz) [reply src/wp-includes/media.php]: Yeah, having `wp_increase_content_media_count()` makes it better, but the count can still be easily manipulated by any theme/plugin. I'll have a better look at the tests to see if this can be avoided.

The output of the `wp_omit_loading_attr_threshold` filter would need to be made into a `static` too to avoid running that filter multiple times. Currently it seems it may run 4-5 tomes per page load.

**Comment by felixarntz** [comment]
@azaozz If we make these variables static, there still has to be a way to reset them. There is no way to make WP's function-based code 100% bulletproof to other developers - if they want to misuse/break it they can already in a million ways.

Your latest commit broke the tests because now the filter isn't re-triggered for new tests. So again, I don't disagree with the approach of making it static, but it doesn't work _just_ like that. There has to be a way to "reset" the filter. I suggest we put it into a function like `wp_omit_loading_attr_threshold( $force = false )`, with the `$force` parameter allowing to enforce re-running it, should the value change, which is necessary for tests to work.

**Comment by azaozz** [comment]
> If we make these variables static, there still has to be a way to reset them.

Hmm, yes but, why? We have to change the functionality just for the sake of testing? Shouldn't be the other way round -- make this work the best way possible then add/fix/tweak the tests to work with the code :)

> Your latest commit broke the tests because...

Yeah, was looking at how to "suggest" changes, i.e. make a variant of the existing patch. However that seems impossible on GH. The only way would be to make another PR which is not good in this case.

Basically I'm thinking that both changes should be made inside `wp_get_loading_attr_default()`. Then it won't need yet another function like `wp_increase_content_media_count()` which doesn't do anything in production and even could be misused, eventually. The only thing stopping this now are the tests that need to be made to work with the changed code.

**Comment by felixarntz** [comment]
@azaozz 
> Hmm, yes but, why? We have to change the functionality just for the sake of testing? Shouldn't be the other way round -- make this work the best way possible then add/fix/tweak the tests to work with the code.

I overall agree with that statement. However, code still has to be written in a testable way if we want to add tests for it - and having state which cannot be controlled in tests prevents that (at a minimum we need to allow reset between tests). We simply cannot "lock" the code so that nobody can override it, as otherwise we also lock it for tests. What we _can_ do is define solid patterns to make it less likely to happen "by accident", and to your point the approach with static variables and dedicated functions makes the access more intentional than just having a global.

> Basically I'm thinking that both changes should be made inside `wp_get_loading_attr_default()`. Then it won't need yet another function like `wp_increase_content_media_count()` which doesn't do anything in production and even could be misused, eventually,  because it can set or reset that count instead of using the filter. The only thing stopping this now are the tests that need to be made to work with the changed code.

See above, I don't think there's a way to make this entirely "locked" static code testable without reducing our test coverage. The reason most WP core tests work out of the box without any of this extra "magic" is that core uses globals, and those can always easily be overwritten, and they automatically get reset between tests. Anywhere where we don't use globals, a special mechanism is needed.

> I'll look at changing/fixing the tests later today and propose another change (well, commit another change as per the above). Then we can see/decide :)

Curious what you come up with. I'll for now update the PR based on your latest commit to make this static mutable again to "unlock" it.

With all that said, the discussion we're having here talks about general code and test patterns in WP core, so I'm not sure this issue is the proper place to have it. Core uses globals throughout, so from that perspective I'd even be in favor of sticking with the approach. If we change this here, it doesn't mean anything, the rest of the codebase will still follow the old pattern, and almost no other core developers will be aware of the new pattern unless it is explicitly called out e.g. in coding standards documentation - it could even be that some folks don't agree with it, another reason not to silo this wide-reaching topic discussion in a very specific PR like this.

**Comment by hellofromtonya** [comment]
Another option for making memoization testable is to move the code into a class and making the memoization value a property. If the class is used as a singleton, the static is not needed; else, it is. That's a design choice. Tests can access the property via reflection or by instantiating a new instance for each test.

Advantages to this approach are:

- the avoidance of added test related code into the production code
- ability to break up larger functions into smaller tasks with private/protected methods

**Comment by azaozz** [comment]
> code still has to be written in a testable way...

Right, the question here is what are we testing. As the `wp_omit_loading_attr_threshold` filter has to run only once per page load (see above), we should probably be testing that (if it makes sense), not running it multiple times which is not supposed to happen.

> We simply cannot "lock" the code so that nobody can override it

Hm, we are not "locking" the code, think of `$content_media_count` as a private property of a class. Functionality is pretty similar. 

The current code as of https://github.com/WordPress/wordpress-develop/pull/1580/commits/1c218d7390ea0fb9ea91939bf7db800c842bb1f2 does not achieve these objectives:
- The `wp_omit_loading_attr_threshold` filter can be run more than once by using the just introduced `wp_omit_loading_attr_threshold( $force = false )` function.
- The `wp_increase_content_media_count( $amount = 1 )` can be used to arbitrarily reset the `$content_media_count` which should not be re-settable/changeable at all.

> With all that said, the discussion we're having here talks about general code and test patterns in WP core...

Hmm, not sure these are "new" things. It's best if tests cover the intended usage, that has been the case for a long time. The current/old tests do not follow the intended usage exactly any more, so they will need adjusting.

The need to change the tests arises from making some changes to the code, namely the `wp_omit_loading_attr_threshold` filter has to run only once per page load, and the `$content_media_count` variable has to be "private" or non-changeable (and it's pretty bad idea to introduce a PHP global at this point).

> ...move the code into a class and making the memoization value a property.
Yes, this would have worked if that code was in a class/wrapper. However changing this is a major refactoring considering back-compat, etc. It may happen one day, but not for 5.9. This patch is a relatively small correction/enhancement to how WordPress adds the `loading` HTML attribute for images and iframes.

**Comment by felixarntz** [comment]
@azaozz 
> we should probably be testing that (if it makes sense), not running it multiple times which is not supposed to happen.

Agreed, we should be testing that as well, this requires adding one more test.

> Hm, we are not "locking" the code, think of `$content_media_count` as a private property of a class. Functionality is pretty similar. 
>
> The current code as of https://github.com/WordPress/wordpress-develop/pull/1580/commits/1c218d7390ea0fb9ea91939bf7db800c842bb1f2 does not achieve these objectives:
> - The `wp_omit_loading_attr_threshold` filter can be run more than once by using the just introduced `wp_omit_loading_attr_threshold( $force = false )` function.
> - The `wp_increase_content_media_count( $amount = 1 )` can be used to arbitrarily reset the `$content_media_count` which should not be re-settable/changeable at all. For example: think of a theme or plugin doing something like `wp_increase_content_media_count( 1000 )`, i.e. disabling the whole thing in the wrong way.

See my previous comment. What you are proposing here makes this code untestable, except if the WP core test suite was starting a new PHP process for every single test which... would make the tests take hours, if not days. I understand this is similar to a private property of a class - to @hellofromtonya's point, that approach would be an appropriate solution, and this is how most sophisticated PHP projects do it. Still, I agree this is too much of a refactoring which this PR is not the place for.

The limitation in the approach you're proposing is that the point of only running once per page-load only applies to that, a page-load. However, PHPUnit tests aren't like a page load, they run tons of things that would normally happen through tons of page loads in a single run.

> Hmm, not sure these are "new" things. It's best if tests cover the intended usage, that has been the case for a long time. The current/old tests do not follow the intended usage exactly any more, so they will need adjusting.

This statement confirms to me that this is not the right place to make this change. We are speaking about a change in coding paradigm that requires refactoring a ton of existing tests and also the related production code - basically "remove globals and manage state elsewhere in a testable way", which is surely not something we can or should arbitrarily decide in an unrelated PR.

**Comment by azaozz** [comment]
> What you are proposing here makes this code untestable

Hmmm, not exactly? Lets look at the tests. The problem with the test for the `wp_omit_loading_attr_threshold` is that it is being initialized more than once with different values. However I don't see what is being tested by initializing it multiple times. Yeah, I understand that using `static` in functions may not be a very well used method, but it seems simplest and fastest. Feel free to implement that in a different way. 

Same for the `$content_media_count` variable. The current/old tests were written to test a global variable. However it was established that it should not be a global variable and the code was changed. So the tests will have to be changed to account for that. A good way would be if the tests do not access that variable at all, as that's how the code is supposed to work now.

> This statement confirms to me that this is not the right place to make this change. We are speaking about a change in coding paradigm that requires refactoring a ton of existing tests...

Uh, sorry but I don't seem to understand exactly what you want to say. What's happening with this patch is: you asked for a review, I did a review, this review established that some changes are needed, the changes were implemented. What's left is to remove the old tests that were written with the old code in mind, and to add new tests that cover the new code. What am I missing?

**Comment by felixarntz** [comment]
@azaozz 
> Hmmm, not exactly? Lets look at the tests. The problem with the test for the `wp_omit_loading_attr_threshold` is that it is being initialized more than once with different values. However I don't see what is being tested by initializing it multiple times.

The filter being applied as expected is being tested. If we can only call it once, we are unable to test that. We could remove that test, but then the production code function could literally be `function() { return 1; }` (i.e. without applying the filter), and the remaining tests wouldn't catch that.

> Uh, sorry but I don't seem to understand exactly what you want to say. What's happening with this patch is: you asked for a review, I did a review, this review established that some changes are needed, the changes were implemented. What's left is to remove the old tests that were written with the old code in mind, and to add new tests that cover the new code. What am I missing?

The tests were already updated to fit the new code, they still test that the filter is applied correctly, and they still pass.

**Comment by azaozz** [comment]
> The filter being applied as expected is being tested. If we can only call it once, we are unable to test that.

Ummm, you mean for a filter to be tested is has to run multiple times? Why? We know the default value and we know the filtered value. running the filter is supposed to change the default to the filtered value. What's the point in running the same test again?

> The tests were already updated to fit the new code..

If I'm not mistaken that update should have included a change to exclude use of the (now supposedly "private")  `$content_media_count` variable. Again, may be missing something here too...

**Comment by felixarntz** [comment]
> Ummm, you mean for a filter to be tested is has to run multiple times? Why? We know the default value and we know the filtered value. Running the filter is supposed to change the default to the filtered value. What's the point in running the same test again?

Technically, if we only run the filter once (e.g. with returning `3`), it could also be that the production function has no filter at all and it just returns `3` hard-coded. This is arguably a bit of nit-picking, but there is another more significant problem: Even if we only ran the filter once in that tests, we would still need to run it again in one of the other tests. If we in any test use the filter to set the number to `3`, it will be `3` throughout all tests, even the ones where we want the default value of the filter to be used. For the content media count, this is even more problematic: There is more than one test that needs to use this functionality, but for any test that is not the very first one being run, it will never start at a count of 0 (since the other test already increased the count). This is why running the filter as well as the content media count variable have to be reset between tests.

There is surely _some_ way to make the tests all pass without a reset, but it would make them hard to understand as well unstable (e.g. what if they run in a different order), since it would pollute the overall state which is a big no-go in tests. Global state needs to be reset between tests so that we mirror production behavior, i.e. ensuring the filter runs once per test which is the test "equivalent" of per page load.

Today, after every core test, all globals are reset, all filters and actions are removed, etc. We have to ensure here as well that the state modified via one test doesn't affect another, that's why a simple `static` without a way to reset won't do it.

**Comment by azaozz** [comment]
> There is surely some way to make the tests all pass without a reset, but it would make them hard to understand...

Right, testing "static caching" is a bit different and somewhat harder. Think of it as testing PHP constants: they need to be tested if they are set, and are set to the proper values, but no other tests are needed as they can never be changed later. So generally they require less testing that variables, etc.

> Today, after every core test, all globals are reset, all filters and actions are removed, etc.

Yes, and the tests here do exactly the same: there are no globals, and the filter is removed. However there is a constant that is set by WordPress when using this functionality. It cannot be reset which is the proper behavior as that's exactly what happens in production.

In any case, you're an experienced developer and know this part of the code very well. We worked really well together when implementing the main functionality. This is a relatively small (by code size) enhancement. Lets add it as-is if you think it's fine.

**Comment by azaozz** [comment]
> That said, it briefly made me think about whether we should actually have a constant here instead of a filter. That probably wouldn't be flexible enough though..

Right, thinking a filter is a lot better there, as it is now.

> We're indeed nit-picking about a rather minor part of the functionality, so we can probably move this forward.

Exactly right 👍

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/52065.
